File: src/rex/sms/MainActivity.java

package rex.sms;

import android.app.Activity;
import android.database.Cursor;
import android.os.Bundle;
import android.provider.Telephony.Sms;
import android.util.Log;

public class MainActivity extends Activity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Cursor c = getContentResolver().query(Sms.Conversations.CONTENT_URI, null, null, null, Sms.Inbox.DEFAULT_SORT_ORDER);
        c.moveToFirst();
        int indexCount = c.getColumnCount();
        while (c.moveToNext()) {
            StringBuilder sb = new StringBuilder();
            sb.append('[');
            for (int i = 0; i < indexCount; i++) {
                sb.append(c.getColumnName(i));
                sb.append("=");
                int type = c.getType(i);
                switch (type) {
                    case Cursor.FIELD_TYPE_FLOAT:
                        sb.append(c.getFloat(i));
                        break;
                    case Cursor.FIELD_TYPE_INTEGER:
                        sb.append(c.getInt(i));
                        break;
                    case Cursor.FIELD_TYPE_STRING:
                        sb.append(c.getString(i));
                        break;
                    default:
                        sb.append("other_types");
                }
                sb.append(",");
            }
            sb.append("]");
            Log.e("XXXXX", sb.toString());
        }
    }

}

File: src/rex/sms/MainActivity.kt
package rex.sms

import android.os.Bundle
import android.support.v4.app.FragmentActivity
import android.util.Log
import com.zy.kotlinutil.permission.permission
import rex.sms.loader.loadThreads

/**
 * Created by zy on 2018/3/22.
 */
class MainActivity : FragmentActivity() {

    private lateinit var slidePresenter: SlidePresenter
    private lateinit var contactPresenter: ContactsListPresenter
    private lateinit var threadContentPresenter: ThreadContentPresenter
    private lateinit var titlePresenter: ThreadTitlePresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        setupViews()

        permission(android.Manifest.permission.READ_SMS, android.Manifest.permission.READ_CONTACTS)
                .onPermitted {
                    Log.e("XXXX", "onPermitted")
                    testCursor()
                }
                .onRefused {
                    Log.e("XXXXX", "refused")
                }
    }

    private fun setupViews() {
        slidePresenter  = SlidePresenter(this)
        contactPresenter = ContactsListPresenter(this)
        threadContentPresenter = ThreadContentPresenter(this)
        titlePresenter = ThreadTitlePresenter(this)

        contactPresenter.onScrollChanged { slidePresenter.expand() }
        contactPresenter.onSelectionChanged { _, smsThread ->
            threadContentPresenter.show(smsThread)
            titlePresenter.show(smsThread)
            slidePresenter.fold()
        }
    }

    fun testCursor() {
        loadThreads { contactPresenter.addThreads(it) }
    }
}